workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/RenderAsState.kt

```diff
@@ -0,0 +1,206 @@
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import okio.ByteString
+
+/**
+ * Runs this [Workflow] as long as this composable is part of the composition, and returns a
+ * [State] object that will be updated whenever the runtime emits a new [RenderingT].
+ *
+ * The workflow runtime will be started when this function is first added to the composition, and
+ * cancelled when it is removed or if the composition fails. The first rendering will be available
+ * immediately as soon as this function returns as [State.value]. Composables that read this value
+ * will automatically recompose whenever the runtime emits a new rendering. In most cases you
+ * probably want to pass the rendering value to the [WorkflowRendering] composable.
+ *
+ * [Snapshot]s from the runtime will automatically be saved and restored using Compose's
+ * [rememberSaveable].
+ *
+ * ## Example
+ *
+ * ```
+ * private val appViewRegistry = ViewRegistry(…)
+ *
+ * @Composable fun App(workflow: Workflow<…>, props: Props) {
+ *   val scaffoldState = …
+ *
+ *   // Run the workflow in the current composition's coroutine scope.
+ *   val rendering by workflow.renderAsState(props, onOutput = { output ->
+ *     // Note that onOutput is a suspend function, so you can run animations
+ *     // and call other suspend functions.
+ *     scaffoldState.snackbarHostState
+ *       .showSnackbar(output.toString())
+ *   })
+ *   val viewEnvironment = remember {
+ *     ViewEnvironment(mapOf(ViewRegistry to appViewRegistry))
+ *   }
+ *
+ *   Scaffold(…) { padding ->
+ *     // Display the root rendering using the view environment's ViewRegistry.
+ *     WorkflowRendering(rendering, viewEnvironment, Modifier.padding(padding))
+ *   }
+ * }
+ * ```
+ *
+ * @receiver The [Workflow] to run. If the value of the receiver changes to a different [Workflow]
+ * while this function is in the composition, the runtime will be restarted with the new workflow.
+ * @param props The [PropsT] for the root [Workflow]. Changes to this value across different
+ * compositions will cause the root workflow to re-render with the new props.
+ * @param interceptors
+ * An optional list of [WorkflowInterceptor]s that will wrap every workflow rendered by the runtime.
+ * Interceptors will be invoked in 0-to-`length` order: the interceptor at index 0 will process the
+ * workflow first, then the interceptor at index 1, etc.
+ * @param scope
+ * The [CoroutineScope] in which to launch the workflow runtime. If not specified, the value of
+ * [rememberCoroutineScope] will be used. Any exceptions thrown in any workflows, after the initial
+ * render pass, will be handled by this scope, and cancelling this scope will cancel the workflow
+ * runtime and any running workers. Note that any dispatcher in this scope will _not_ be used to
+ * execute the very first render pass.
```

Comment:
which will instead be executed on the ... main dispatcher (?) ...?

Reply:
I think it'll be executed by whatever is running the caller, no?

Reply:
Yep. I'll make that explicit.